okio/src/appleMain/kotlin/okio/BufferedSource.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+import kotlinx.cinterop.CPointer
+import kotlinx.cinterop.CPointerVar
+import kotlinx.cinterop.UnsafeNumber
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.convert
+import kotlinx.cinterop.pointed
+import kotlinx.cinterop.reinterpret
+import kotlinx.cinterop.usePinned
+import kotlinx.cinterop.value
+import platform.Foundation.NSData
+import platform.Foundation.NSError
+import platform.Foundation.NSInputStream
+import platform.Foundation.NSLocalizedDescriptionKey
+import platform.Foundation.NSUnderlyingErrorKey
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.darwin.NSUIntegerVar
+import platform.posix.memcpy
+import platform.posix.uint8_tVar
+
+fun BufferedSource.inputStream(): NSInputStream = BufferedSourceInputStream(this)
+
+/** Returns an input stream that reads from this source. */
+@OptIn(UnsafeNumber::class)
+private class BufferedSourceInputStream(
+  private val bufferedSource: BufferedSource
+) : NSInputStream(NSData()) {
+
+  private var error: NSError? = null
+
+  override fun streamError(): NSError? = error
+
+  override fun open() {
+    // no-op
+  }
+
+  override fun read(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+    try {
+      val internalBuffer = bufferedSource.buffer
+
+      if (bufferedSource is RealBufferedSource) {
+        if (bufferedSource.closed) throw IOException("closed")
+
+        if (internalBuffer.size == 0L) {
```

Comment:
Please rewrite this if clause as  

```
if (bufferedSource.exhausted()) return 0
```

That will guarantee the buffer has at least one byte in it after!

Reply:
Oh, nice. I see `RealBufferedSource.commonExhausted()` is this exact logic. I went ahead and replaced this same logic in other places as well https://github.com/square/okio/pull/1123/commits/3571012c6ab78f1c41865afe2c763b9369727499. Let me know if you'd rather have that reverted though.